File: gaussian.jl
# Cris' function rewritten
# Creates a half Gaussian kernel with the given parameters.
function MakeHalfGaussian(sigma, derivativeOrder)
    # Compute the size of the kernel and adjust it if necessary
    halfFilterSize = 1 + 9

    # Initialize the filter array and some variables
    filter = zeros(halfFilterSize)
    r0 = halfFilterSize
    sigma2 = sigma * sigma

    # Compute the filter values based on the derivative order
    # ignores constant factors
    # does not cycle over central values
    if derivativeOrder == 0
        factor = -0.5 / sigma2
        normalization = 0.0
        filter[r0] = 1.0  # central value is 1 since exp(0) = 1
        for rr in 1:halfFilterSize-1
            rad = Float64(rr)
            g = exp(factor * rad^2)
            filter[r0-rr] = g
            normalization += g
        end
        normalization = 1.0 / (normalization * 2 + 1) # central value added, and account for both tails
        for rr in 1:halfFilterSize
            filter[rr] *= normalization
        end

    elseif derivativeOrder == 1
        factor = -0.5 / sigma2
        moment = 0.0
        filter[r0] = 0.0  # central value must be 0 since x as prefactor in 1st derivative
        for rr in 1:halfFilterSize-1
            rad = Float64(rr)
            g = rad * exp(factor * rad^2)
            filter[r0-rr] = g
            moment += rad * g
        end
        normalization = 1.0 / (2.0 * moment)  # moment in centre is 0
        for rr in 1:halfFilterSize-1 # centre needs no normalisation since it is 0
            filter[rr] *= normalization
        end

    elseif derivativeOrder == 2
        norm = 1.0 / (sqrt(2.0 * pi) * sigma * sigma2)
        mean = 0.0
        filter[r0] = -norm
        for rr in 1:halfFilterSize-1
            rad = Float64(rr)
            sr2 = rad^2 / sigma2
            g = (sr2 - 1.0) * norm * exp(-0.5 * sr2)
            filter[r0-rr] = g
            mean += g
        end
        mean = (mean * 2.0 + filter[r0]) / (Float64(r0) * 2.0 + 1.0)
        filter[r0] -= mean
        moment = 0.0
        for rr in 1:halfFilterSize-1
            rad = Float64(rr)
            filter[r0-rr] -= mean
            moment += rad^2 * filter[r0-rr]
        end
        normalization = 1.0 / moment
        for rr in 1:halfFilterSize
            filter[rr] *= normalization
        end
    elseif derivativeOrder == 3
        norm = 1.0 / (sqrt(2.0 * pi) * sigma * sigma2^2)
        filter[r0] = 0.0
        moment = 0.0
        for rr in 1:halfFilterSize-1
            rad = Float64(rr)
            rr2 = rad^2
            sr2 = rr2 / sigma2
            g = norm * exp(-0.5 * sr2) * (rad * (3.0 - sr2))
            filter[r0-rr] = g
            moment += g * rr2 * rad
        end
        normalization = 3.0 / moment
        for rr in 0:halfFilterSize-1
            filter[rr] *= normalization
        end
    else
        error("Not implemented")
    end
    
    # Return the filter array
    return filter
end

function makeGaussJH(sigma,width)
    N(x) = 1/(sqrt(2*pi)*sigma) * exp(-x^2/(2*sigma))
    dN(x) = -(x/sigma^2) * N(x)
    ddN(x) = -(1/sigma^2) * N(x) - (x/sigma^2) * dN(x)

    #x = collect(-width:width)
    x = ImageFiltering.OffsetVector(-width:width,-(width+1))
    #ddg = ddN.(x)
    g = KernelFactors.gaussian(sigma,2*width+1)
    g_dd = g .* ((x.^2 .- sigma^2) ./ (sigma^4))
    
    E = mean(g_dd)
    g_dd .= g_dd .- E

    m2 = dot(g_dd[-width:-1],(-width:-1).^2)
    g_dd .= g_dd ./ (m2)

    return g_dd
end


function makeGauss2(sigma,width)
    N(x) = 1/(sqrt(2*pi)*sigma) * exp(-x^2/(2*sigma))
    dN(x) = -(x/sigma^2) * N(x)
    ddN(x) = -(1/sigma^2) * N(x) - (x/sigma^2) * dN(x)

    x = collect(-width:width)
    g_dd = ddN.(x)
    g_dd = ImageFiltering.OffsetVector(g_dd,-(width+1))
    
    E = mean(g_dd)
    g_dd .= g_dd .- E  # assure zero result in convolution of constant

    #m2 = dot(g_dd[-width:-1],(-width:-1).^2)
    #g_dd .= g_dd ./ (m2)
    m2 = dot(g_dd,x.^2)
    g_dd .= 2*g_dd ./ (m2)

    return g_dd
end

dd = makeGaussJH(3,9)
w = div(length(dd), 2)
x = collect(-w:w)

dot(dd,ones(length(dd)))
dot(dd,x)
dot(dd,x.^2)

File: src/gaussian.py
import numpy as np
import math
import matplotlib.pyplot as plt

g = make_half_gaussian(3,1)
dg = np.concatenate([-g,np.flip(g[0:-1])])
g = make_half_gaussian(3,2)
ddg = np.concatenate([g,np.flip(g[0:-1])])

w = len(ddg) // 2
x = np.array(range(-w,w+1))

np.dot(dg,np.ones(len(dg)))
np.dot(dg,x)
np.dot(dg,np.power(x,2))

np.dot(ddg,np.ones(len(ddg)))
np.dot(ddg,x)
np.dot(ddg,np.power(x,2))



plt.plot(ddg)
plt.show()


def make_half_gaussian(sigma, derivative_order):
	half_gaussian_size = 9
	half_filter_size = 1 + half_gaussian_size

	filter = np.zeros(half_filter_size)
	r0 = half_filter_size - 1
	sigma2 = sigma * sigma

	if derivative_order == 0:
		factor = -0.5 / sigma2
		normalization = 0
		filter[r0] = 1.0
		for rr in range(1, half_filter_size):
			rad = float(rr)
			g = math.exp(factor * (rad * rad))
			filter[r0 - rr] = g
			normalization += g
		normalization = 1.0 / (normalization * 2 + 1)
		for rr in range(half_filter_size):
			filter[rr] *= normalization

	elif derivative_order == 1:
		factor = -0.5 / sigma2
		moment = 0.0
		filter[r0] = 0.0
		for rr in range(1, half_filter_size):
			rad = float(rr)
			g = rad * math.exp(factor * (rad * rad))
			filter[r0 - rr] = g
			moment += rad * g
		normalization = 1.0 / (2.0 * moment)
		for rr in range(half_filter_size - 1):
			filter[rr] *= normalization

	elif derivative_order == 2:
		norm = 1.0 / (math.sqrt(2.0 * math.pi) * sigma * sigma2)
		mean = 0.0
		filter[r0] = -norm
		for rr in range(1, half_filter_size):
			rad = float(rr)
			sr2 = rad * rad / sigma2
			g = (sr2 - 1.0) * norm * math.exp(-0.5 * sr2)
			filter[r0 - rr] = g
			mean += g
		mean = (mean * 2.0 + filter[r0]) / (float(r0) * 2.0 + 1.0)  # mean of all values in filter
		filter[r0] -= mean
		moment = 0.0
		for rr in range(1, half_filter_size):
			rad = float(rr)
			filter[r0 - rr] -= mean
			moment += rad * rad * filter[r0 - rr]
		normalization = 1.0 / moment
		for rr in range(half_filter_size):
			filter[rr] *= normalization

	elif derivative_order == 3:
		norm = 1.0 / (math.sqrt(2.0 * math.pi) * sigma * sigma2 * sigma2)
		filter[r0] = 0.0
		moment = 0.0
		for rr in range(1, half_filter_size):
			rad = float(rr)
			rr2 = rad * rad
			sr2 = rr2 / sigma2
			g = norm * math.exp(-0.5 * sr2) * (rad * (3.0 - sr2))
			filter[r0 - rr] = g
			moment += g * rr2 * rad
		normalization = 3.0 / moment
		for rr in range(half_filter_size):
			filter[rr] *= normalization

	else:
		raise NotImplementedError("Derivative order not implemented")

	return filter
